src/emulator/renderer/src/texture_format.cpp

```diff
@@ -149,4 +171,306 @@ void tiled_texture_to_linear_texture(uint8_t *dest, const uint8_t *src, uint16_t
+    switch (base_format) {
+    case SCE_GXM_TEXTURE_BASE_FORMAT_UBC1:
+    case SCE_GXM_TEXTURE_BASE_FORMAT_UBC2:
+    case SCE_GXM_TEXTURE_BASE_FORMAT_UBC3:
+        return ((width + 3) / 4) * ((height + 3) / 4) * ((base_format == SCE_GXM_TEXTURE_BASE_FORMAT_UBC1) ? 8 : 16);
+    default:
+        break;
+    }
+
+    return 0;
+}
+
+// =========================== COMPRESSION ============================
+// Some texture has block compression, when uncompressed will have swizzled layout. Since on some backend, no
+// option is provided to make the GPU driver not try to translate the layout to linear, we have to do uncompress
+// and unswizzled on the CPU.
+
+// This is a modified version of Benjamin Dobell's DXT decompression for compatible with our codebase and future OS.
+
+/**
+ * \brief Helper method that packs RGBA channels into a single 4 byte pixel, but reversed for little endian.
+ * 
+ * \param r   red channel.
+ * \param g   green channel.
+ * \param b   blue channel.
+ * \param a   alpha channel.
+ */
+static std::uint32_t pack_rgba_reversed(std::uint8_t r, std::uint8_t g, std::uint8_t b, std::uint8_t a) {
+    return ((a << 24) | (b << 16) | (g << 8) | r);
+}
+
+/**
+ * \brief Decompresses one block of a DXT1 texture and stores the resulting pixels at the appropriate offset in 'image',
+ *        with custom alpha table.
+ * 
+ * \param x                 x-coordinate of the first pixel in the block.
+ * \param y                 y-coordinate of the first pixel in the block.
+ * \param width             width of the texture being decompressed.
+ * \param height            height of the texture being decompressed.
+ * \param block_storage     pointer to the block to decompress.
+ * \param image             pointer to image where the decompressed pixel data should be stored.
+ * \param alpha_table       alpha lookup table.
+ **/
+static void decompress_block_dxt1_shared(std::uint32_t x, std::uint32_t y, std::uint32_t width, const std::uint8_t *block_storage, std::uint32_t *image,
+    const std::uint8_t *alpha_table) {
+    std::uint16_t color0 = *reinterpret_cast<const std::uint16_t *>(block_storage);
+    std::uint16_t color1 = *reinterpret_cast<const std::uint16_t *>(block_storage + 2);
+
+    std::uint32_t temp;
+
+    temp = (color0 >> 11) * 255 + 16;
+    std::uint8_t r0 = (std::uint8_t)((temp / 32 + temp) / 32);
+    temp = ((color0 & 0x07E0) >> 5) * 255 + 32;
+    std::uint8_t g0 = (std::uint8_t)((temp / 64 + temp) / 64);
+    temp = (color0 & 0x001F) * 255 + 16;
+    std::uint8_t b0 = (std::uint8_t)((temp / 32 + temp) / 32);
+
+    temp = (color1 >> 11) * 255 + 16;
+    std::uint8_t r1 = (std::uint8_t)((temp / 32 + temp) / 32);
+    temp = ((color1 & 0x07E0) >> 5) * 255 + 32;
+    std::uint8_t g1 = (std::uint8_t)((temp / 64 + temp) / 64);
+    temp = (color1 & 0x001F) * 255 + 16;
+    std::uint8_t b1 = (std::uint8_t)((temp / 32 + temp) / 32);
```

Comment:
I thought you would use `static_cast` but I guess it is fine.

Reply:
This code is copy from internet. If you want I will adjust

Reply:
I figured. It is up to you if you want to change.

But if you don't change that means the c-style casts have won.

Reply:
Sorry  i hope you dont mind so much >~<. I promised to be get back to that, not for now